### DISCUSSION TITLE
Inconsistent hiding of wiimotes

### DISCUSSION THREAD

**Discussion by Trihy** [nefarius/HidHide#85] [Q&A]
- [x ] I have searched open and closed issues for duplicates
- [ x] I am submitting a bug report for existing functionality that does not work as intended
- [x ] I have read https://vigem.org/Community-Support/
- [x ] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
Hiding wiimotes gives inconsistent results.

### Steps to reproduce
-Just connect wiimotes with Wiimotehook or similar tool, could be dolphin emulator too.
-Use hidhide to hide them.
-Works fine.

-But once they disconnect and reconnects, they can be seen again on windows joy.cpl

Hidhide still show them as hided. But they are not.

**Actual result:Sometimes are hided, sometimes are not
**Expected result:Stay hided

### Machine info

**CPU-Architecture:x64
**Windows version: Windows 11
**Software/driver version(s):HidHide Drivers v1.2.98

**Comment by Kanuan** [comment]
When the issue happens, are you just reconnecting the device or are you re-pairing it to windows?

## Edit 1

To clarify, repairing means: 

- making windows itself search and ask the device for connection

OR

- When that notification "a device is attempting to connect" appears and you choose to accept it

**Reply by Trihy** [reply]
Hi. No, not re-pairing, at least not with the pairing button. Just connecting after a while or after a system restart. 

Since Windows 10, wiimotes cannot be directly paired. At least the TR ones, not sure about the non TR ones (so you never see windows asking for connection permission or any notification related)

So you need to use wiimotehook or other program to pair them correctly. Dolphin emulator can do that too. But to reconnect (if you turn it off) you need to press a button or sometimes the 1+2 buttons, which can be somehow considered re-pairing? Maybe Windows assign another id when this happens.

**Reply by Kanuan** [reply]
Yeah... Things got complicated quickly.

Basically, there IS a bug with HidHide in which, when re-pairing a device, HidHide Configuration Client will SHOW it is the same device (and still "hidden" if it was before), but it will have a new Identification at system level and therefore won't be actually hidden.

What I think is happening when using this wiimotehook software is that it is not keeping the same device identification, either because of a bug, they didn't care since it doesn't make much of a difference, they can't because some windows stuff or some other explanation.

If the Device identification is different then HidHide will treat it as a new device and won't make it hidden. Unfortunately this is per-design and won't be changed since hiding devices only by its name has lead to many issues in the HidGuardian era.

**Reply by Trihy** [reply]
I think that can be related to that too. 

To re-hide it I have to disable/enable hide device a couple of times on hidhide. Since it doesnt show as a different device.

Maybe any workaround to hide by name on this kind of devices? Wiimotes are way more tricky than ps controllers.

**Reply by Kanuan** [reply]
> Only different thing is that when gets another id, hidhide doesnt show another device, show the same device hided, but not really hided. Is that correct too?

Yeah, that is a bug. HidHide should either show a new device or, if it somehow knows it is the same device but with a different ID then mark it as un-hidden.

But what is actually happening with HidHide not actually hiding the controller since it's a different ID is per-design

**Reply by Trihy** [reply]
Thanks. Well, at least learned a little more how hidhide works. 

I´d edited the previous post but guess those possible workarounds are not planned atm.

**Comment by Trihy** [comment]
Steam after september update, can pair wiimotes. Well, will do it without asking. So another problem becomes when not hiding them. Dolphin and other programs wont work anymore with wiimotes if not closing steam first. Steam games are a no go, since steam takes control over wiimote.

Wonder if another workaround can be added to handle this.

**Reply by nefarius** [reply]
I can't make fixes for hardware I do not own if owners of said hardware don't contribute fixes or insights. Until then this will remain permanently stale.

**Reply by Trihy** [reply]
Hi. Think main problem is when wiimotes resync, they get another id which confuses hidhide and they re appear. Don't think it's a bug, just the way hidhide works. 

I was reading about your tool to rewrite hw id but not sure if will help on this case.

**Comment by Trihy** [comment]
Did hidhide ever tried to hide devices by mac? Is it possible? Or viable?

**Reply by nefarius** [reply]
Nope, not every device has a MAC address so that is not possible to do consistently.

**Comment by Trihy** [comment]
Hi. Do you think we can add some exceptions for wiimotes? Since lightgun games are having a comeback in the community. And not being able to hide wiimotes is causing many problems. I have both TR and non TR wiimotes and could help with testing :)

**Reply by nefarius** [reply]
Nope, out of scope. Feel free to fork and submit a fix yourself or make a tool that uses the API or CLI. We do not have the time not consider this something that should be in the main package.

**Reply by Trihy** [reply]
Thanks for your answer.

Do you think will be easier or better to block by vid/pid or by Mac address?

**Reply by nefarius** [reply]
If we are talking about HidHide's core feature; you can not block by either of those without additional apps or scripts, HidHide blocks devices identified **by Instance ID only** (e.g. this `HID\VID_054C&PID_0CE6&MI_03\A&26D6E322&0&0000`). You need to craft your own custom tools to get from VID/PID or MAC to the Instance ID, then it will work.

**Reply by Trihy** [reply]
Thanks. Yes, I know it blocks by ID. Wiimotes use to get another ID everytime you hit 1+2 or sync button. But will try to implement other ways. I think vid&pid could be easier to do than mac address. Since I dont have too much knowledge.

**Reply by nefarius** [reply]
The ID changes more often on wireless devices because Windows has no clue how to make it truly unique if the device doesn't e.g. report a unique serial number in a standard way that Windows understands. So it's pretty much the fault of lazy hardware designers, otherwise it would always remain the same ID and you wouldn't have this problem.

**Reply by Trihy** [reply]
Well, after all, Wiimotes weren't designed to be used on Windows, so can't blame them :D

Will take a read at hidhide source and see if I can implement a vid&pid hide. 

Thanks for the info and support.

**Reply by nefarius** [reply]
> Well, after all, Wiimotes weren't designed to be used on Windows, so can't blame them :D

You wouldn't believe how many folks don't think that far and complain why peripherals not designed for a certain platform do not work as convenient for them 😉

**Reply by Trihy** [reply]
Been playing a little with CLI, to see better how hidhide works.

Im not sure if hiding HID will be enough or hiding BTHENUM will be needed too.

Lets see how it goes.

non TR and TR have a little different PID

--dev-hide "HID\{00001124-0000-1000-8000-00805f9b34fb}_VID&0002057e_PID&0306\8&1f03f731&10&0000"
--dev-hide "HID\{00001124-0000-1000-8000-00805f9b34fb}_VID&0002057e_PID&0330\8&1fdd1ee3&4&0000"

Everything after the latest \ I guess is the instance ID
